### PR TITLE
SE-510 [LX-129] REST API to get tag data for LabXchange

### PR DESCRIPTION
**Pull request by SSPJ**
## Description

This PR concerns the creation of an API which the LabXchange frontend can query for the purpose of displaying information about items (e.g. "subject area", "subtitles language", "depth of knowledge", "license")

* Can query a list of entities at api/v1/entities/
* Tags are at api/v1/entities/<entity>/tags/
* Can be filtered by taxonomy eg tags/?taxonomy=7

It is read-only and does not have auth.

## Author Comments, Concerns, and Open Questions

One issue which came up was the use of urls for `external_id` in entities. That seemed a reasonable use case to me, but it presents an extremely challenging (to me) edge case, because many web servers do not permit url encoding of forward slashes in the path of the url, at least by default. Such that if the `external_id` is `https://url.org`, I do not know how it is possible to tell Django that you mean `api/v1/entities/https:%2F%2Furl.org/tags`.

If it's important to support urls, I will think on this issue further.

## Test Instructions

1. Create some taxonomies and tags
1. Create some entities and tag them
1. On this branch, start the server and curl the endpoints

All entities are shown:
```
$ curl -s http://0.0.0.0:18250/api/v1/entities/ | jq
[
  {
    "id": 3,
    "entity_type": "url",
    "external_id": "https://somethingsomething.com"
  },
  {
    "id": 1,
    "entity_type": "xblock",
    "external_id": "block-v1:OpenCraft+B+1+type@drag-and-drop-v2+block@971aadaaaa5a4886b37fee3ced711b30"
  }
]
```

A single entity is shown:
```
$ curl -s http://0.0.0.0:18250/api/v1/entities/block-v1:OpenCraft+B+1+type@drag-and-drop-v2+block@971aadaaaa5a4886b37fee3ced711b30/ | jq
{
  "id": 1,
  "entity_type": "xblock",
  "external_id": "block-v1:OpenCraft+B+1+type@drag-and-drop-v2+block@971aadaaaa5a4886b37fee3ced711b30"
}
```

All tags are shown:
```
$ curl -s http://0.0.0.0:18250/api/v1/entities/block-v1:OpenCraft+B+1+type@drag-and-drop-v2+block@971aadaaaa5a4886b37fee3ced711b30/tags/ | jq
{
  "tags": [
    {
      "taxonomy_uid": 7,
      "taxonomy_name": "Subject Area",
      "tag": "Biochemistry"
    },
    {
      "taxonomy_uid": 9,
      "taxonomy_name": "License",
      "tag": "CC-BY-SA-4.0"
    },
    {
      "taxonomy_uid": 7,
      "taxonomy_name": "Subject Area",
      "tag": "Cell biology"
    },
    {
      "taxonomy_uid": 10,
      "taxonomy_name": "Subtitles Language",
      "tag": "汉语"
    }
  ]
}
```

Filtered tags are shown:
```
$ curl -s http://0.0.0.0:18250/api/v1/entities/block-v1:OpenCraft+B+1+type@drag-and-drop-v2+block@971aadaaaa5a4886b37fee3ced711b30/tags/?taxonomies=7,9 | jq
{
  "tags": [
    {
      "taxonomy_uid": 7,
      "taxonomy_name": "Subject Area",
      "tag": "Biochemistry"
    },
    {
      "taxonomy_uid": 9,
      "taxonomy_name": "License",
      "tag": "CC-BY-SA-4.0"
    },
    {
      "taxonomy_uid": 7,
      "taxonomy_name": "Subject Area",
      "tag": "Cell biology"
    }
  ]
}
```

The API is read-only:
```
$ curl -X PUT -d type=things -d external_id=stuff http://0.0.0.0:18250/api/v1/entities/
```

## TODOs
If anything isn't yet done, list it here
- [x] Squash before merging

### CODE REVIEW

**Comment by bradenmacdonald** [comment]
Re URLs, I don't think we need to support them but it should be possible to have django accept them "normally"/"unencoded", so the URL would look like `/api/v1/entities/https://url.org/tags` (with a urlpattern like `/api/v1/entities/.*/tags/$`). I don't think that's a great format, nor that we should use it, but it should work.

File: blockstore/apps/api/v1/tests/test_views.py [hunk 1]
```diff
@@ -51,6 +54,27 @@ def setUp(self):
...
+                name='tag2',
+                taxonomy=self.taxonomy,
+            ),
+        ]
+
+        self.entity = EntityFactory(
+            entity_type='xblock',
+            external_id='some-resource-uri',
+        )
+        self.entity.tags = self.tags
```
Comment (bradenmacdonald) [inline blockstore/apps/api/v1/tests/test_views.py]: What is the benefit of factory methods over the normal API?
```
self.tagstore = DjangoTagstore()

self.taxonomy = self.tagstore.create_taxonomy("This Taxonomy")
self.tags = [
    self.taxonomy.add_tag('tag1'),
    self.taxonomy.add_tag('tag2'),
]

self.entity = EntityId(entity_type='xblock', external_id='some-resource-uri')
for t in self.tags:
    self.tagstore.add_tag_to(t, self.entity)
```

File: blockstore/apps/api/v1/views/entities.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        else:
+            queryset = entity.tags.all()
+
+        extracted = [self._convert(tag) for tag in queryset]
+        serializer = EntityTagSerializer({'tags': extracted})
+        return Response(serializer.data)
+
+    def retrieve(self, request, pk=None, external_id=None):  # pylint: disable=unused-argument
+        '''
+        Get a single tag belonging to an entity.
```
Comment (bradenmacdonald) [inline blockstore/apps/api/v1/views/entities.py]: Is this view really that useful?

Comment (SSPJ) [reply blockstore/apps/api/v1/views/entities.py]: Not especially.

**Comment by SSPJ** [comment]
> @SSPJ For now, as much as possible, I'd like to keep the tag-related code separate from the blockstore code, so could you please move the REST/API code into a django app like tagstore/tagstore_rest or something instead of blockstore/apps/api ?

Yep, have done so.

> This approach bypasses the generic Tagstore API and uses the django tagstore backend API directly. Did you consider using APIView to implement this API in a way that's not tied to the django model backend, or would that be too much work? We are only using the django backend so it's probably fine either way, but it's nice to not close doors if we don't have to.

@bradenmacdonald I think it depends how strongly you feel about it. It's definitely do-able. The Tagstore API right now doesn't have the right methods to avoid a lot of redundant database calls. (In particular, there's no way, given an entity, to get the names of every taxonomy its tags are in, but also we can't get an entity by it's external_id or filter entity tags by taxonomy.) I'm happy to extend the Tagstore API, but I do worry about rewriting functionality that is already well-tested in Django itself.

Let me know!

**Comment by bradenmacdonald** [comment]
@SSPJ I would say it's fine to stick with the django implementation but please don't expose any django-specific details via the REST API (i.e. identify tags by their `name` not by their database ID, don't expose the database ID of `entity` objects, etc. which [is an edX API best practice anyways](https://openedx.atlassian.net/wiki/spaces/OpenDev/pages/18350757/edX+REST+API+Conventions#edXRESTAPIConventions-2.IdentifyingResources)).

File: blockstore/apps/api/v1/tests/test_views.py (hunk 1 above)
Comment (SSPJ) [reply blockstore/apps/api/v1/tests/test_views.py]: No benefits that I know. I picked it for consistency with the other tests in blockstore and that's a moot point as it's no longer in blockstore. Thanks for the example code :)

**Comment by SSPJ** [comment]
@bradenmacdonald I've made the changes you've suggested (with the exception of the tagstore api for the entity lookup, as we discussed). Do you see anything else that needs addressed here?

File: tagstore/tagstore_rest/v1/urls.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+from ..routers import DefaultRouter
+from .views.entities import EntityViewSet, EntityTagViewSet
+
+root_router = DefaultRouter()
+
+root_router.register(r'entities', EntityViewSet)
+
+entities_nested_router = nested_routers.NestedDefaultRouter(root_router, r'entities', lookup='')
+entities_nested_router.register(r'tags', EntityTagViewSet, base_name='entitytags')
```
[the remark this answers is not in the document]
Comment (SSPJ) [reply tagstore/tagstore_rest/v1/urls.py]: Very concise!

File: tagstore/tagstore_rest/v1/views/entities.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+from ..serializers.entities import EntitySerializer, EntityTagSerializer
+
+
+class EntityViewSet(viewsets.ReadOnlyModelViewSet):
+    '''
+    ViewSet for Entity model.
+    '''
+
+    lookup_field = 'external_id'
```
[the remark this answers is not in the document]
Comment (SSPJ) [reply tagstore/tagstore_rest/v1/views/entities.py]: `get_lookup_regex` is about accepting comma separated lists of params. I took out the nested routers to implement the type/external_id look up via explicitly declared routes.

**Comment by SSPJ** [comment]
@bradenmacdonald The new url looks like (for example) http://0.0.0.0:18250/tagstore/api/v1/entities/xblock/block-v1:OpenCraft+B+1+type@drag-and-drop-v2+block@971aadaaaa5a4886b37fee3ced711b30/tags?taxonomies=7,9 . Does everything look okay?

Would you like me to deal with the merge conflict or would you rather?

**Comment by bradenmacdonald** [comment]
:+1: Thanks @SSPJ ! Conditional approval if you change the URL patterns so that it works with a trailing slash, for consistency with the rest of the API.

e.g.
http://localhost:18250/tagstore/api/v1/entities/thing/dandelion/tags/ not http://localhost:18250/tagstore/api/v1/entities/thing/dandelion/tags

http://localhost:18250/tagstore/api/v1/entities/thing/dandelion/ not http://localhost:18250/tagstore/api/v1/entities/thing/dandelion

- [x] I tested this: tested the API via the web
- [x] I read through the code
- [x] I checked for accessibility issues
- [x] Includes documentation

One that's fixed, you can please merge. The easiest way is not to rebase but rather to merge master into this branch (`git fetch origin --prune && get merge origin/master -m "Update with master"`). Then once you resolve the comments, you can push to GitHub then merge using GitHub's "Squash and merge" feature. (Or regular merge if you think the commit history is important.)

**Comment by pomegranited** [comment]
@SSPJ Do you mind to update this PR to show the final API URLs for the tagstore API you added?  I'm having to dig through the code to find them for LX-130, and would prefer to find PRs that are self-documenting 😄   Thanks!

**Comment by SSPJ** [comment]
Of course, @pomegranited . Sorry about that!

**Comment by pomegranited** [comment]
No worries at all @SSPJ :)  Thanks for updating the PR!

**Comment by edx-webhook** [comment]
Although this pull request is already merged, I've created [OSPR-5470](https://openedx.atlassian.net/browse/OSPR-5470) so that we can track it in Jira.

There is nothing you have to do. No action is needed from your side.  Thanks again for your contribution.

**Pull request merged**